File: core/event_bus.py
from typing import Dict, List, Callable
from loguru import logger

class EventBus:
    def __init__(self):
        self._subscribers: Dict[str, List[Callable]] = {}
        logger.info("EventBus initialized")

    def subscribe(self, event_type: str, callback: Callable):
        if event_type not in self._subscribers:
            self._subscribers[event_type] = []
        self._subscribers[event_type].append(callback)
        logger.debug(f"Subscribed to event: {event_type}")

    def publish(self, event_type: str, data: dict):
        if event_type in self._subscribers:
            for callback in self._subscribers[event_type]:
                try:
                    callback(data)
                except Exception as e:
                    logger.error(f"Error in event handler: {e}")
        logger.debug(f"Published event: {event_type}")

File: project/src/core/event-bus.ts
import { logger } from './logger';

type Callback = (data: any) => void;

export class EventBus {
  private subscribers: Map<string, Callback[]>;

  constructor() {
    this.subscribers = new Map();
    logger.info('EventBus initialized');
  }

  subscribe(eventType: string, callback: Callback): void {
    const subscribers = this.subscribers.get(eventType) || [];
    subscribers.push(callback);
    this.subscribers.set(eventType, subscribers);
    logger.debug({ eventType }, 'Subscribed to event');
  }

  publish(eventType: string, data: any): void {
    const subscribers = this.subscribers.get(eventType) || [];
    subscribers.forEach(callback => {
      try {
        callback(data);
      } catch (error) {
        logger.error({ error, eventType }, 'Error in event handler');
      }
    });
    logger.debug({ eventType }, 'Published event');
  }
}
